pgx/src/datum/into.rs

```diff
@@ -32,6 +33,73 @@ pub trait IntoDatum {
         unsafe { pg_sys::get_array_type(Self::type_oid()) }
     }
+
+    /// Is a Datum of this type compatible with another Postgres type?
+    ///
+    /// An example of this are the Postgres `text` and `varchar` types, which are both
+    /// technically compatible from a Rust type perspective.  They're both represented in Rust as
+    /// `String` (or `&str`), but the underlying Postgres types are different.
+    ///
+    /// If implementing this yourself, you likely want to follow a pattern like this:
+    ///
+    /// ```rust,no_run
+    /// # use pgx::*;
+    /// # #[repr(transparent)]
+    /// # struct FooType(String);
+    /// # impl pgx::IntoDatum for FooType {      
+    ///     fn is_compatible_with(other: pg_sys::Oid) -> bool {
+    ///         // first, if our type is the other type, then we're compatible
+    ///         Self::type_oid() == other
+    ///
+    ///         // and here's the other type we're compatible with
+    ///         || other == pg_sys::VARCHAROID
+    ///     }
+    ///
+    /// #    fn into_datum(self) -> Option<pg_sys::Datum> {
+    /// #        todo!()
+    /// #    }
+    /// #
+    /// #    fn type_oid() -> pg_sys::Oid {
+    /// #        pg_sys::TEXTOID
+    /// #    }
+    /// # }
+    /// ```
+    #[inline]
+    fn is_compatible_with(other: pg_sys::Oid) -> bool {
+        Self::type_oid() == other
+    }
+
+    /// Is a Datum of this type pass by value or pass by reference?
+    ///
+    /// We provide a hardcoded list of known Postgres types that are pass by value,
+    /// but you are free to implement this yourself for custom types.
+    #[inline]
+    fn is_pass_by_value() -> bool
+    where
+        Self: 'static,
+    {
+        let my_type = std::any::TypeId::of::<Self>();
+        my_type == std::any::TypeId::of::<i8>()
+            || my_type == std::any::TypeId::of::<i16>()
+            || my_type == std::any::TypeId::of::<i32>()
+            || my_type == std::any::TypeId::of::<i64>()
+            || my_type == std::any::TypeId::of::<u8>()
+            || my_type == std::any::TypeId::of::<u16>()
+            || my_type == std::any::TypeId::of::<u32>()
+            || my_type == std::any::TypeId::of::<u64>()
+            || my_type == std::any::TypeId::of::<f32>()
+            || my_type == std::any::TypeId::of::<f64>()
+            || my_type == std::any::TypeId::of::<bool>()
+            || my_type == std::any::TypeId::of::<()>()
+            || my_type == std::any::TypeId::of::<crate::Time>()
+            || my_type == std::any::TypeId::of::<crate::TimeWithTimeZone>()
+            || my_type == std::any::TypeId::of::<crate::Timestamp>()
+            || my_type == std::any::TypeId::of::<crate::TimestampWithTimeZone>()
+            || my_type == std::any::TypeId::of::<crate::Date>()
+            || my_type == std::any::TypeId::of::<PgOid>()
+            || my_type == std::any::TypeId::of::<pg_sys::Datum>()
+            || my_type == std::any::TypeId::of::<Option<pg_sys::Datum>>()
```

Comment:
Wait. Sorry.

I re-skimmed this and this line is incorrect. Part of my audit PR makes it correct, but this line should go until then.

```suggestion
```

Alternatively, this is also true, but I suspect was not what you meant:
```suggestion
            || my_type == std::any::TypeId::of::<Option<&pg_sys::Datum>>()
```

Reply:
I don't understand what you're asking me to do on this one, or why what I had is wrong or why making it a reference to a `pg_sys::Datum` would instead be correct?  What am I missing?

Reply:
So my understanding is that currently, Datum has a repr of a usize, and only things that fit within a usize are "pass by value". An Option requires has a tag bit to distinguish between Some and None, which means that on a 64-bit target, `Option<usize>` requires 65 bits. There is a trick called the "null pointer optimization", but that only applies to pointers, not numbers. Any integer value is valid, some pointers are not. `&usize` only has to occupy as much space as a pointer does, so even though `&usize` and `usize` are both 64 bits, `Option<&usize>` still only requires 64 bits.

Reply:
Ahhh I discussed this and turns out I was just confused.